### PR TITLE
Merge annotations before propagation

### LINKED ISSUE #752: EL Reconciler keeps updating Deployment
# Expected Behavior

Reconcile succeeds eventually.

# Actual Behavior

In each reconcile cycle, we update the deployment and this keeps on going forever

# Steps to Reproduce the Problem

1. Install Triggers v0.8.0
2. Create a new EL e.g. `examples/gitlab`
3. Look at EL logs

# Additional Info

Snip:

```
{"level":"error","logger":"controller","caller":"controller/controller.go:449","msg":"Reconcile error","error":"Operation cannot be fulfilled on deployments.apps \"el-gitlab-listener\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"knative.dev/pkg/controller.(*Impl).handleErr\n\tknative.dev/pkg@v0.0.0-20200625173728-dfb81cf04a7c/controller/controller.go:449\nknative.dev/pkg/controller.(*Impl).processNextWorkItem\n\tknative.dev/pkg@v0.0.0-20200625173728-dfb81cf04a7c/controller/controller.go:435\nknative.dev/pkg/controller.(*Impl).RunContext.func2\n\tknative.dev/pkg@v0.0.0-20200625173728-dfb81cf04a7c/controller/controller.go:372"}
{"level":"info","logger":"controller","caller":"controller/controller.go:436","msg":"Reconcile failed. Time taken: 196.495284ms","knative.dev/traceid":"5fcd1f02-9620-4939-beed-1e4a5ad3fef6","knative.dev/key":"default/gitlab-listener"}
{"level":"info","logger":"controller","caller":"eventlistener/eventlistener.go:409","msg":"Updated EventListener Deployment el-gitlab-listener in Namespace default","knative.dev/traceid":"a92deaef-213f-4407-80e3-3c37f69ce582","knative.dev/key":"default/gitlab-listener"}
{"level":"info","logger":"controller","caller":"controller/controller.go:443","msg":"Reconcile succeeded. Time taken: 205.315364ms","knative.dev/traceid":"a92deaef-213f-4407-80e3-3c37f69ce582","knative.dev/key":"default/gitlab-listener"}
{"level":"error","logger":"controller","caller":"eventlistener/eventlistener.go:406","msg":"Error updating EventListener Deployment: Operation cannot be fulfilled on deployments.apps \"el-gitlab-listener\": the object has been modified; please apply your changes to the latest version and try again","knative.dev/traceid":"32abd542-c7b9-4083-9c9d-204da16a92bd","knative.dev/key":"default/gitlab-listener","stacktrace":"github.com/tektoncd/triggers/pkg/reconciler/v1alpha1/eventlistener.(*Reconciler).reconcileDeployment\n\tgithub.com/tektoncd/triggers/pkg/reconciler/v1alpha1/eventlistener/eventlistener.go:406\ngithub.com/tektoncd/triggers/pkg/reconciler/v1alpha1/eventlistener.(*Reconciler).ReconcileKind\n\tgithub.com/tektoncd/triggers/pkg/reconciler/v1alpha1/eventlistener/eventlistener.go:124\ngithub.com/tektoncd/triggers/pkg/client/injection/reconciler/triggers/v1alpha1/eventlistener.(*reconcilerImpl).Reconcile\n\tgithub.com/tektoncd/triggers/pkg/client/injection/reconciler/triggers/v1alpha1/eventlistener/reconciler.go:169\nknative.dev/pkg/controller.(*Impl).processNextWorkItem\n\tknative.dev/pkg@v0.0.0-20200625173728-dfb81cf04a7c/controller/controller.go:434\nknative.dev/pkg/controller.(*Impl).RunContext.func2\n\tknative.dev/pkg@v0.0.0-20200625173728-dfb81cf04a7c/controller/controller.go:372"}
{"level":"error","logger":"controller","caller":"eventlistener/reconciler.go:210","msg":"Returned an error","knative.dev/traceid":"32abd542-c7b9-4083-9c9d-204da16a92bd","knative.dev/key":"default/gitlab-listener","targetMethod":"ReconcileKind","error":"Operation cannot be fulfilled on deployments.apps \"el-gitlab-listener\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/tektoncd/triggers/pkg/client/injection/reconciler/triggers/v1alpha1/eventlistener.(*reconcilerImpl).Reconcile\n\tgithub.com/tektoncd/triggers/pkg/client/injection/reconciler/triggers/v1alpha1/eventlistener/reconciler.go:210\nknative.dev/pkg/controller.(*Impl).processNextWorkItem\n\tknative.dev/pkg@v0.0.0-20200625173728-dfb81cf04a7c/controller/controller.go:434\nknative.dev/pkg/controller.(*Impl).RunContext.func2\n\tknative.dev/pkg@v0.0.0-20200625173728-dfb81cf04a7c/controller/controller.go:372"}
{"level":"error","logger":"controller","caller":"controller/controller.go:449","msg":"Reconcile error","error":"Operation cannot be fulfilled on deployments.apps \"el-gitlab-listener\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"knative.dev/pkg/controller.(*Impl).handleErr\n\tknative.dev/pkg@v0.0.0-20200625173728-dfb81cf04a7c/controller/controller.go:449\nknative.dev/pkg/controller.(*Impl).processNextWorkItem\n\tknative.dev/pkg@v0.0.0-20200625173728-dfb81cf04a7c/controller/controller.go:435\nknative.dev/pkg/controller.(*Impl).RunContext.func2\n\tknative.dev/pkg@v0.0.0-20200625173728-dfb81cf04a7c/controller/controller.go:372"}
{"level":"info","logger":"controller.event-broadcaster","caller":"record/event.go:274","msg":"Event(v1.ObjectReference{Kind:\"EventListener\", Namespace:\"default\", Name:\"gitlab-listener\", UID:\"ef6ce08f-68d4-40b3-9bf2-6ca063b3a9aa\", APIVersion:\"triggers.tekton.dev/v1alpha1\", ResourceVersion:\"31450256\", FieldPath:\"\"}): type: 'Warning' reason: 'InternalError' Operation cannot be fulfilled on deployments.apps \"el-gitlab-listener\": the object has been modified; please apply your changes to the latest version and try again"}
{"level":"info","logger":"controller","caller":"controller/controller.go:436","msg":"Reconcile failed. Time taken: 198.373314ms","knative.dev/traceid":"32abd542-c7b9-4083-9c9d-204da16a92bd","knative.dev/key":"default/gitlab-listener"}
{"level":"info","logger":"controller","caller":"eventlistener/eventlistener.go:409","msg":"Updated EventListener Deployment el-gitlab-listener in Namespace default","knative.dev/traceid":"0baf82e9-dc18-4b18-bed7-d711e5ebef4b","knative.dev/key":"default/gitlab-listener"}
{"level":"info","logger":"controller","caller":"controller/controller.go:443","msg":"Reconcile succeeded. Time taken: 195.897503ms","knative.dev/traceid":"0baf82e9-dc18-4b18-bed7-d711e5ebef4b","knative.dev/key":"default/gitlab-listener"}


```

### PR DESCRIPTION
**Pull request by dibyom**
In #712, we added a feature to propagate annotations added to the EL to its
underlying resources. However, this resulted in infinite reconcile loops since
the deployment controller will add a standard revision annotation that the EL
controller will keep trying to overwrite. To fix this, this commit switches the
annotation propagation to merge any annotations set on the underlying resources
before adding any extra annotations from the EL.

Fixes #752

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes
```release-note
     Triggers no longer overwrites annotations set on the underlying deployment and service objects. 
```

### CODE REVIEW

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener_test.go [hunk 1]
```diff
@@ -608,6 +604,7 @@ func TestReconcile(t *testing.T) {
 			ConfigMaps:     []*corev1.ConfigMap{loggingConfigMap},
 		},
 	}, {
+		// Check that if a user manually updates the annotations for a service, we do not revert the change.
```
Comment (savitaashture) [inline pkg/reconciler/v1alpha1/eventlistener/eventlistener_test.go:607]:
```suggestion
		// Check that if a user manually updates the annotations for a deployment, we do not revert the change.
```

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener.go [hunk 2]
```diff
@@ -153,18 +153,21 @@ func (r *Reconciler) FinalizeKind(ctx context.Context, el *v1alpha1.EventListene
...
-	if !reflect.DeepEqual(oldMeta.Labels, newMeta.Labels) {
+func reconcileObjectMeta(existing *metav1.ObjectMeta, desired metav1.ObjectMeta) (updated bool) {
+	if !reflect.DeepEqual(existing.Labels, desired.Labels) {
 		updated = true
-		oldMeta.Labels = newMeta.Labels
+		existing.Labels = desired.Labels
 	}
-	if !reflect.DeepEqual(oldMeta.Annotations, newMeta.Annotations) {
+
+	// TODO(dibyom): We should propagation of some labels such as `kubernetes.io/last-applied-revision`
```
Comment (savitaashture) [inline pkg/reconciler/v1alpha1/eventlistener/eventlistener.go:162]:
```suggestion
	// TODO(dibyom): We should propagation of some annotations such as `kubernetes.io/last-applied-revision`
```
is that annotations or labels ?

**Review by savitaashture — COMMENTED** [review]
minor nit on code comments

otherwise
/lgtm

Thank you

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener.go (hunk 2 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/eventlistener/eventlistener.go]: done!

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener_test.go (hunk 1 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/eventlistener/eventlistener_test.go]: done

**Comment by dibyom** [comment]
> 2020/09/11 14:31:06 main.go:319: Something went wrong: failed to prepare test environment: --provider=gke boskos failed to acquire project: Post "http://boskos.test-pods.svc.cluster.local./acquire?dest=busy&owner=pull-tekton-triggers-integration-tests&request_id=fc368477-d314-46cf-9e05-82596a6c306c&state=free&type=gke-project": dial tcp 10.55.241.233:80: connect: connection refused

Looks like a test infra issue

**Review by khrm — APPROVED** [review]

**Pull request merged**